## Publications
List of reference papers in Computational Biology. 

This is a list of publications (or other links) that I use as reference in my research. 


## A. Academic publications
#### 1A. Bioconductor:  
Two main academic papers describing Bioconductor 
 
 * 1.1 Original description written when the project was initiated:  
 Gentleman RC, Carey VJ, Bates DM, et al. Bioconductor: open software development for computational biology and bioinformatics.
 Genome Biology. [2004](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC545600/) ;5(10):R80.  
 
 * 1.2 Update. Future prospects in Bioconductor:  
 Huber et al. Orchestrating high-throughput genomic analysis with Bioconductor. 
 Nature Methods. [2015](http://www.nature.com/nmeth/journal/v12/n2/abs/nmeth.3252.html); 12, 115–121.  
 

#### 2A. GRanges / IRanges
* Lawrence M, Huber W, Pagès H, Aboyoun P, Carlson M, Gentleman R, et al. ([2013](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003118)) Software for Computing and Annotating Genomic Ranges. PLoS Comput Biol 9(8): e1003118.  

#### 3A. Reproducible Research
* Peng. Reproducible Research in Computational Science. Science. [2011](http://www.sciencemag.org/content/334/6060/1226.abstract); 334(6060) 1226-1227  


#### 4A. Batch effects  
* Leek JT, Scharpf RB, Corrada Bravo H, Simcha D, Langmead B, Johnson WE, Geman D, Baggerly Km, Irizarry RA. Tackling the widespread and critical impact of batch effects in high-throughput data. Nature Reviews Genetics 11, ([2010](http://www.nature.com/nrg/journal/v11/n10/full/nrg2825.html)) 733-739  


#### 5A. DNA algorithms
* 5.1 Boyer-Moore: the benchmark for exact matching algorithm.  
 Boyer RS, Moore JS. A fast string searching algorithm. Communications of the ACM, (1977) 762-772 doi:10.1145/359842.359859 https://www.cs.utexas.edu/~moore/publications/fstrpos.pdf  
* 5.2 Li L, Homer N. A survey of sequence alignment algorithms for next-generation sequencing. Brief. Bioinform. 11, (2010) 473-483 http://www.ncbi.nlm.nih.gov/pubmed/20460430  
* Gusfield, D. Algorithms on Strings, Trees and Sequences: Computer Science and Computational Biology. ISBN-13: 978-0521585194 http://www.amazon.com/Algorithms-Strings-Trees-Sequences-Computational/dp/0521585198
  
<br>  
  
## B. Reference Links
  * [Free Science Data Books](http://www.wzchen.com/data-science-books) by William Chen [@wzchen](https://twitter.com/wzchen)  
  * [Bioconductor for Genomic Data Science](http://kasperdanielhansen.github.io/genbioconductor/) by Kasper Daniel Hansen [@kasperdhansen](https://twitter.com/kasperdhansen)  
  * Data Analysis for Genomics
This is based on the the PH525 series by [Rafael A Irizarry](https://twitter.com/rafalab), [Michael Love](https://twitter.com/mikelove) and the the PH525x team from Dana Farber & Harvard School Public Health. http://genomicsclass.github.io/book/  
  * Plots to avoid:** http://genomicsclass.github.io/book/pages/plots_to_avoid.html
  
<br>  
  


  
## C. Books  
Some important books available online for free:

* **R for Data Science** by Hadley Wickham. Get the book [here](https://www.amazon.com/R-Data-Science-Hadley-Wickham/dp/1491910399/ref=as_li_ss_tl?ie=UTF8&qid=1469550189&sr=8-1&keywords=R+for+data+science&linkCode=sl1&tag=devtools-20&linkId=6fe0069f9605cf847ed96c191f4e84dd) or read it for [free](http://r4ds.had.co.nz)  
* **Advanced R** by Hadley Wickham. Get the book [here](https://www.amazon.com/dp/1466586966/ref=cm_sw_su_dp?tag=devtools-20) or read it for [free](http://adv-r.had.co.nz/). 
* **R packages** by Hadley Wickham. Get the book [here](https://www.amazon.com/dp/1491910593/ref=cm_sw_su_dp?tag=r-pkgs-20) or read it for [free](http://r-pkgs.had.co.nz/).   
* **Happy Git and GitHub for the useR** by Jenny Bryan. Read the [book](http://happygitwithr.com/).  
* **Text Mining with R** by Julia Silge and David Robinson. Read the [book](https://www.tidytextmining.com/) . 

<br>  


## D. R packages
  * h-index over years. My first R package. Plotting the [h-index](https://github.com/jdieramon/hIndex) over years for a given author.    
  * [Swirl Courses](https://github.com/swirldev/swirl_courses#swirl-courses)




